doc/SAI-Proposal-HW-Resource.md

```diff
@@ -0,0 +1,183 @@
+    By default this trap is created in 
+    SAI_SWITCH_ATTR_DEFAULT_TRAP_GROUP trap group.
+or
+2. create_hostif_trap():
+    For trap group created in step 1.
+3. create_policer()
+4. Attach policer to the trap using set of 
+    SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER attribute.
+```
+
+### Sample New Workflow
+-------------------
+New workflow will identify if a given trap is applicable to ingress/egress/both. This will be done based on the enhanced capability query API.
+```
+1. Query the HW capability for a given trap
+    sai_attribute_t trap_attra[2];
+    trap_attra[0].id = 
+        SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+    trap_attrs[0].value.s32 = 
+        SAI_HOSTIF_TRAP_TYPE_DNAT_MISS;
+    
+    trap_attrs[2].id = 
+        SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+    trap_attrs[2].value.s32 = 
+        SAI_HOSTIF_TRAP_TYPE_NAT_HAIRPIN;  
+    
+    sai_status_t sai_query_object_stage(
+        switch_id,
+        SAI_OBJECT_TYPE_HOSTIF_TRAP,
+        2,
+        trap_attrs,
+        stage);
+
+    SAI adapter will fill buffer stage with appropriate HW pipeline stage where this trap is handled. As an example, a given HW implementation MAY return buffer stage filled as following.
+    stage[0] = SAI_OBJECT_STAGE_INGRESS
+    stage[1] = SAI_OBJECT_STAGE_INGRESS
+
+2. Create a policer pool for ingress stage
+    sai_object_id_t ingress_policer_id;
+    sai_attribute_t ingress_policer_attrs[2];
+
+    ingress_policer_attrs[0].id =
+        (sai_attr_id_t)SAI_POLICER_ATTR_METER_TYPE;
+    ingress_policer_attrs[0].value =
+        SAI_METER_TYPE_PACKETS;
+
+    ingress_policer_attrs[1].id =
+        (sai_attr_id_t)SAI_POLICER_ATTR_OBJECT_STAGE;
+    ingress_policer_attrs[1].value =
+        SAI_OBJECT_STAGE_INGRESS;
+
+    sai_hostif_trap_api->create_policer(
+        &ingress_policer_id,
+        2,
+        ingress_policer_attrs);
+
+3. Create a hostif trap group for ingress stage
+    sai_object_id_t trap_group_id;
+    sai_attribute_t trap_group_attrs[3];
+
+    trap_group_attrs[0].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_GROUP_ATTR_QUEUE;
+    trap_group_attrs[0].value.u32 =
+        0x10;
+
+    trap_group_attrs[1].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER,
+    trap_group_attrs[1].value.oid =
+        ingress_policer_id;
+
+    trap_group_attrs[2].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_GROUP_ATTR_OBJECT_STAGE,
+    trap_group_attrs[2].value =
+        SAI_OBJECT_STAGE_INGRESS;
+
+    sai_hostif_trap_group_api->create_hostif_trap_group(
+        &trap_group_id,
+        3,
+        trap_group_attrs);
+
+4. Create hostif trap for ingress stage
+    sai_object_id_t trap_id;
+    sai_attribute_t trap_attrs[3];
+
+    trap_attrs[0].id =
+        (sai_attr_id_t) SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+    trap_attrs[0].value =
+        SAI_HOSTIF_TRAP_TYPE_DNAT_MISS;
+        
+    trap_attrs[1].id =
+        (sai_attr_id_t) SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+    trap_attrs[1].value =
+        SAI_HOSTIF_TRAP_TYPE_NAT_HAIRPIN;

```

Comment:
We cannot use the same attribute-id twice for a given create API call. Please remove lines 162 to 165

Reply:
https://github.com/opencomputeproject/SAI/commit/85e155118b61485ae3060f6e3b249b6388810810